tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1553 @@
+
+    // Read in the raw data for all entries
+    uint8_t* entryBuffer = (uint8_t*)tsk_malloc(metadataEntriesSize);
+    if (entryBuffer == nullptr) {
+        writeError("BitlockerParser::readFveMetadataEntries: Error allocating memory for entries");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    size_t bytesRead = tsk_img_read(m_img_info, currentOffset, (char*)entryBuffer, metadataEntriesSize);
+    if (bytesRead != metadataEntriesSize) {
+        writeError("BitlockerParser::readFveMetadataBlockHeader: Error reading metadata entries (read " + to_string(bytesRead) + " bytes");
+        free(entryBuffer);
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Parse the metadata entries
+    list<string> errorList;
+    if (BITLOCKER_STATUS::SUCCESS != readMetadataEntries(entryBuffer, metadataEntriesSize, m_metadataEntries)) {
+        free(entryBuffer);
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    free(entryBuffer);
+    return BITLOCKER_STATUS::SUCCESS;
+}
+
+/**
+* Get the volume master key.
+* General idea is that we're trying to find and parse a volume master key (VMK) entry to get the key
+* we will use to decrypt the full volume encryption key. There are frequently more than one of these entries present.
+* For example, we might have a password protected entry and a recovery password protected entry. 
+* 
+* On success, m_decryptedVmkEntry will contain the decrypted key to use in the next step.
+* 
+* @return SUCCESS if we successfully decrypted the volume master key
+*         GENERAL_ERROR if an unspecified error occurs
+*         WRONG_PASSWORD if we found a password/recovery password protected VMK but the password we have was incorrect
+*         NEEDS_PASSWORD if we found a password/recovery password protected VMK but do not have a password
+*         UNSUPPORTED_KEY_PROTECTION_TYPE if we found a VMK entry with a key protected by an unsupported method
+*/
+BITLOCKER_STATUS BitlockerParser::getVolumeMasterKey() {
+    m_decryptedVmkEntry = nullptr;
+
+    // Get VMK entries
+    list<MetadataEntry*> vmkEntries;
+    getMetadataEntries(m_metadataEntries, BITLOCKER_METADATA_ENTRY_TYPE::VOLUME_MASTER_KEY, BITLOCKER_METADATA_VALUE_TYPE::VOLUME_MASTER_KEY, vmkEntries);
+    if (vmkEntries.empty()) {
+        writeError("BitlockerParser::setVolumeMasterKey: No Volume Master Key entries found");
+        return BITLOCKER_STATUS::GENERAL_ERROR;
+    }
+
+    // Attempt to parse each of the VMK entries, keeping track of some specific errors if they occur.
+    BITLOCKER_STATUS ret = BITLOCKER_STATUS::GENERAL_ERROR;
+    MetadataEntry* vmk = nullptr;
+    bool possibleMissingPassword = false;
+    bool possibleWrongPassword = false;
+    bool possibleUnsupportedProtectionType = false;
+    for (auto it = vmkEntries.begin(); it != vmkEntries.end(); ++it) {
+        ret = parseVMKEntry(*it, &vmk);
+        if (ret == BITLOCKER_STATUS::SUCCESS && vmk != nullptr) {
```

Comment:
I think we should be freeing vmk at the end of this loop?

Reply:
So I think we don't need to because vmk will never be allocated if we failed and on success we save it to be freed later. I think it just needs better comments.